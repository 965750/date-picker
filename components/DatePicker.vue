<template>
  <div>
    <div class="datePicker__header">
      <div class="datePicker__arrowIcon" @click="changeMonth(-1)">
        <v-icon name="chevron-left" @click="changeMonth(-1)" />
      </div>
      <p>{{ month | namesOfMonths }} {{ year }}</p>
      <div class="datePicker__arrowIcon" @click="changeMonth(1)">
        <v-icon name="chevron-right" />
      </div>
    </div>
    <div class="datePicker__calendar">
      <div class="datePicker__daysLabels">
        <div
          v-for="day in days"
          :key="day"
          class="datePicker__dayLabel"
        >
          {{ day | shortLabel }}
        </div>
      </div>
      <div class="datePicker__days">
        <div
          v-for="space in daysSpacing"
          :key="`spacing-${space}`"
          class="day"
        />
        <div
          v-for="day in daysInMonth"
          :key="`calendar-${day.label}`"
          :class="[`day`, day.availableType === 1 ? 'day__wrapper--available' : '']"
        >
          <div
            :class="[
              day.availableType === 2 ? 'day__wrapper--circle' : '',
              day.availableType === 3 ? 'day__wrapper--circle' : '',
            ]"
          />

          <div
            v-if="day.availableType === 3 || day.availableType === 2"
            :class="[
              day.availableType === 2 ? 'day__wrapper--availableRight' : '',
              day.availableType === 3 ? 'day__wrapper--availableLeft' : '',
            ]"
          />
          <span
            :class="[
              'day__label',
              day.availableType === 2 || day.availableType === 3 ? 'day__label--available' : '',
            ]"
          >
            {{ day.label }}
          </span>
        </div>
      </div>
    </div>
    {{ month }} --- {{ year }}
  </div>
</template>

<script>
export default {
  name: 'DatePicker',
  props: {
    availableDates: {
      required: true,
      type: Array
    }
  },
  data () {
    return {
      days: ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'],
      year: 2019,
      month: 3
    }
  },
  computed: {
    daysSpacing () {
      const spacing = new Date(this.year, this.month, 0).getDay() + 1

      if (spacing > 6) {
        return 0
      } else {
        return spacing
      }
    },
    daysInMonth () {
      const days = new Array(new Date(this.year, this.month + 1, 0).getDate())
      days.fill(1)

      const fullDays = days.map((day, index) => {
        const timestamp = new Date(this.year, this.month, index + 1).getTime()
        let availableType = 0

        this.availableDates.forEach((date) => {
          if (timestamp < date.end.getTime() && timestamp > date.start.getTime()) {
            availableType = 1
          }

          if (timestamp === date.start.getTime()) {
            availableType = 2
          }

          if (timestamp === date.end.getTime()) {
            availableType = 3
          }
        })

        return {
          label: index + 1,
          timestamp,
          availableType
        }
      })
      console.log(fullDays, 44444)
      return fullDays
    }
  },
  methods: {
    changeMonth (value) {
      console.log('clicked')
      if (this.month === 11 && value === 1) {
        this.year++
        this.month = 0
      } else if (this.month === 0 && value === -1) {
        this.year--
        this.month = 11
      } else {
        this.month += value
      }
    }
  }
}
</script>

<style lang="scss">
.datePicker {

  &__header {
    background: #00DBB1;
    height: 60px;
    line-height: 60px;
    color: #FFF;
    text-align: center;
    font-size: 20px;
    font-weight: bold;
    display: flex;
    justify-content: space-between;
    padding: 0 25px;
  }

  &__arrowIcon {
    padding: 0 5px;
    color: #44514D;
    cursor: pointer;
  }

  &__days {
    margin-top: 20px;
    display: flex;
    flex-wrap: wrap;
  }

  &__daysLabels {
    display: flex;
    font-size: 14px;
    color: #C7C9CC;
  }

  &__dayLabel {
    width: 40px;
    text-align: center;
  }

  &__calendar {
    padding: 19px;
    border: 1px solid #D6D6D6;
    min-height: 260px;
  }
}

.day {
  text-align: center;
  width: 40px;
  height: 40px;
  line-height: 40px;
  position: relative;

  &:hover {
    border: 2px solid #01DBB3;
    border-radius: 50%;
    cursor: pointer;
  }

  &__label {
    z-index: 3;
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);

    &--available {
      color: #FFF;
    }
  }
  &__wrapper {

    &--available {
      background: #C3FEF8;
      color: #3AD8B8;

      &:hover {
        border: 2px solid #01DBB3;
        border-radius: 50%;
        cursor: pointer;
      }
    }

    &--circle {
      position: absolute;
      top: 0;
      z-index: 2;
      width: 100%;
      height: 100%;
      border-radius: 50%;
      background: #01DBB3;
      color: #FFF;
    }

    &--availableRight {
      position: absolute;
      z-index: 1;
      background: #C3FEF8;
      width: 50%;
      height: 100%;
      top: 0;
      right: 0;
    }

    &--availableLeft {
      @extend .day__wrapper--availableRight;
      left: 0;
    }
  }
}
</style>
