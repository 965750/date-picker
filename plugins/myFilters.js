import Vue from 'vue'

export function shortLabel (value) {
  return `${value.substr(0, 3)}`
}
const filters = { shortLabel }

Object.keys(filters).forEach((key) => {
  Vue.filter(key, filters[key])
})
