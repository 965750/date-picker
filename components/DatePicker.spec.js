import {
  mount,
  createLocalVue
} from '@vue/test-utils'
import { shortLabel, namesOfMonths } from '../plugins/myFilters'
import DatePicker from './DatePicker'

const localVue = createLocalVue()

localVue.filter('shortLabel', shortLabel)
localVue.filter('namesOfMonths', namesOfMonths)

const initialProps = {
  calendarType: 'start',
  availableDates: [
    {
      start: new Date(2019, 8, 3),
      end: new Date(2019, 8, 10)
    },
    {
      start: new Date(2019, 8, 22),
      end: new Date(2019, 8, 27)
    },
    {
      start: new Date(2019, 9, 8),
      end: new Date(2019, 9, 24)
    }
  ]
}

describe('DatePicker.vue', () => {
  const createWrapper = (props = initialProps) => mount(DatePicker, {
    localVue,
    propsData: {
      ...props
    },
    stubs: {
      VIcon: true
    }
  })

  it('should match snapshot', () => {
    const wrapper = createWrapper()

    expect(wrapper.element).toMatchSnapshot()
  })

  it('should move calendar to previous month', () => {
    const wrapper = createWrapper()

    wrapper.setData({
      year: 2017,
      month: 3
    })

    expect(wrapper.find('#calendarHeader').text()).toBe('April 2017')

    wrapper.find('#arrowLeft').trigger('click')

    expect(wrapper.find('#calendarHeader').text()).toBe('March 2017')
  })

  it('should move calendar to previous month and year', () => {
    const wrapper = createWrapper()

    wrapper.setData({
      year: 2017,
      month: 0
    })

    expect(wrapper.find('#calendarHeader').text()).toBe('January 2017')

    wrapper.find('#arrowLeft').trigger('click')

    expect(wrapper.find('#calendarHeader').text()).toBe('December 2016')
  })

  it('should move calendar to next month and year', () => {
    const wrapper = createWrapper()

    wrapper.setData({
      year: 2017,
      month: 11
    })

    expect(wrapper.find('#calendarHeader').text()).toBe('December 2017')

    wrapper.find('#arrowRight').trigger('click')

    expect(wrapper.find('#calendarHeader').text()).toBe('January 2018')
  })

  it('should emit changeDate with day that was clicked', () => {
    const wrapper = createWrapper()

    const day = {
      label: 9,
      timestamp: 1567980000000
    }

    wrapper.vm.pickDate(day)

    expect(wrapper.emitted().changeDate[0][0]).toBe(day)
  })

  it('should return true when day from calendar match day picked', () => {
    const wrapper = createWrapper()

    const day = {
      label: 9,
      timestamp: 1567980000000
    }

    wrapper.setData({
      pickedDates: {
        start: day,
        end: null
      }
    })

    expect(wrapper.vm.isPicked(day)).toBe(true)
  })

  it('should return false when day from calendar does NOT match day picked', () => {
    const wrapper = createWrapper()

    const day = {
      label: 9,
      timestamp: 1567980000000
    }

    wrapper.setData({
      pickedDates: {
        end: {
          label: 1,
          timestamp: 1567288800000
        },
        start: null
      }
    })

    expect(wrapper.vm.isPicked(day)).toBe(false)
  })
})
