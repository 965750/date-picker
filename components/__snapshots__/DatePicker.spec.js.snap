// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatePicker.vue should match snapshot 1`] = `
<div>
  <div
    class="datePicker__header"
  >
    <div
      class="datePicker__arrowIcon"
      id="arrowLeft"
    >
      <vicon-stub
        name="chevron-left"
      />
    </div>
     
    <p
      id="calendarHeader"
    >
      
      September 2019
    
    </p>
     
    <div
      class="datePicker__arrowIcon"
      id="arrowRight"
    >
      <vicon-stub
        name="chevron-right"
      />
    </div>
  </div>
   
  <div
    class="datePicker__calendar"
  >
    <div
      class="datePicker__daysLabels"
    >
      <div
        class="datePicker__dayLabel"
      >
        
        Sun
      
      </div>
      <div
        class="datePicker__dayLabel"
      >
        
        Mon
      
      </div>
      <div
        class="datePicker__dayLabel"
      >
        
        Tue
      
      </div>
      <div
        class="datePicker__dayLabel"
      >
        
        Wed
      
      </div>
      <div
        class="datePicker__dayLabel"
      >
        
        Thu
      
      </div>
      <div
        class="datePicker__dayLabel"
      >
        
        Fri
      
      </div>
      <div
        class="datePicker__dayLabel"
      >
        
        Sat
      
      </div>
    </div>
     
    <div
      class="datePicker__days"
    >
       
      <div
        class="day"
      >
        <div
          class=""
        />
         
        <!---->
         
        <span
          class="day__label"
        >
          
          1
        
        </span>
      </div>
      <div
        class="day"
      >
        <div
          class=""
        />
         
        <!---->
         
        <span
          class="day__label"
        >
          
          2
        
        </span>
      </div>
      <div
        class="day"
      >
        <div
          class="day__wrapper--circle"
        />
         
        <div
          class="day__wrapper--availableRight"
        />
         
        <span
          class="day__label day__label--available"
        >
          
          3
        
        </span>
      </div>
      <div
        class="day day__wrapper--available"
      >
        <div
          class=""
        />
         
        <!---->
         
        <span
          class="day__label"
        >
          
          4
        
        </span>
      </div>
      <div
        class="day day__wrapper--available"
      >
        <div
          class=""
        />
         
        <!---->
         
        <span
          class="day__label"
        >
          
          5
        
        </span>
      </div>
      <div
        class="day day__wrapper--available"
      >
        <div
          class=""
        />
         
        <!---->
         
        <span
          class="day__label"
        >
          
          6
        
        </span>
      </div>
      <div
        class="day day__wrapper--available"
      >
        <div
          class=""
        />
         
        <!---->
         
        <span
          class="day__label"
        >
          
          7
        
        </span>
      </div>
      <div
        class="day day__wrapper--available"
      >
        <div
          class=""
        />
         
        <!---->
         
        <span
          class="day__label"
        >
          
          8
        
        </span>
      </div>
      <div
        class="day day__wrapper--available"
      >
        <div
          class=""
        />
         
        <!---->
         
        <span
          class="day__label"
        >
          
          9
        
        </span>
      </div>
      <div
        class="day"
      >
        <div
          class="day__wrapper--circle"
        />
         
        <div
          class="day__wrapper--availableLeft"
        />
         
        <span
          class="day__label day__label--available"
        >
          
          10
        
        </span>
      </div>
      <div
        class="day"
      >
        <div
          class=""
        />
         
        <!---->
         
        <span
          class="day__label"
        >
          
          11
        
        </span>
      </div>
      <div
        class="day"
      >
        <div
          class=""
        />
         
        <!---->
         
        <span
          class="day__label"
        >
          
          12
        
        </span>
      </div>
      <div
        class="day"
      >
        <div
          class=""
        />
         
        <!---->
         
        <span
          class="day__label"
        >
          
          13
        
        </span>
      </div>
      <div
        class="day"
      >
        <div
          class=""
        />
         
        <!---->
         
        <span
          class="day__label"
        >
          
          14
        
        </span>
      </div>
      <div
        class="day"
      >
        <div
          class=""
        />
         
        <!---->
         
        <span
          class="day__label"
        >
          
          15
        
        </span>
      </div>
      <div
        class="day"
      >
        <div
          class=""
        />
         
        <!---->
         
        <span
          class="day__label"
        >
          
          16
        
        </span>
      </div>
      <div
        class="day"
      >
        <div
          class=""
        />
         
        <!---->
         
        <span
          class="day__label"
        >
          
          17
        
        </span>
      </div>
      <div
        class="day"
      >
        <div
          class=""
        />
         
        <!---->
         
        <span
          class="day__label"
        >
          
          18
        
        </span>
      </div>
      <div
        class="day"
      >
        <div
          class=""
        />
         
        <!---->
         
        <span
          class="day__label"
        >
          
          19
        
        </span>
      </div>
      <div
        class="day"
      >
        <div
          class=""
        />
         
        <!---->
         
        <span
          class="day__label"
        >
          
          20
        
        </span>
      </div>
      <div
        class="day"
      >
        <div
          class=""
        />
         
        <!---->
         
        <span
          class="day__label"
        >
          
          21
        
        </span>
      </div>
      <div
        class="day"
      >
        <div
          class="day__wrapper--circle"
        />
         
        <div
          class="day__wrapper--availableRight"
        />
         
        <span
          class="day__label day__label--available"
        >
          
          22
        
        </span>
      </div>
      <div
        class="day day__wrapper--available"
      >
        <div
          class=""
        />
         
        <!---->
         
        <span
          class="day__label"
        >
          
          23
        
        </span>
      </div>
      <div
        class="day day__wrapper--available"
      >
        <div
          class=""
        />
         
        <!---->
         
        <span
          class="day__label"
        >
          
          24
        
        </span>
      </div>
      <div
        class="day day__wrapper--available"
      >
        <div
          class=""
        />
         
        <!---->
         
        <span
          class="day__label"
        >
          
          25
        
        </span>
      </div>
      <div
        class="day day__wrapper--available"
      >
        <div
          class=""
        />
         
        <!---->
         
        <span
          class="day__label"
        >
          
          26
        
        </span>
      </div>
      <div
        class="day"
      >
        <div
          class="day__wrapper--circle"
        />
         
        <div
          class="day__wrapper--availableLeft"
        />
         
        <span
          class="day__label day__label--available"
        >
          
          27
        
        </span>
      </div>
      <div
        class="day"
      >
        <div
          class=""
        />
         
        <!---->
         
        <span
          class="day__label"
        >
          
          28
        
        </span>
      </div>
      <div
        class="day"
      >
        <div
          class=""
        />
         
        <!---->
         
        <span
          class="day__label"
        >
          
          29
        
        </span>
      </div>
      <div
        class="day"
      >
        <div
          class=""
        />
         
        <!---->
         
        <span
          class="day__label"
        >
          
          30
        
        </span>
      </div>
    </div>
  </div>
   
  <!---->
</div>
`;
